pkg/middleware/rate_limit_test.go

```diff
@@ -0,0 +1,81 @@
+package middleware
+
+import (
+	"net/http"
+	"net/http/httptest"
+	"testing"
+	"time"
+
+	"github.com/grafana/grafana/pkg/models"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"gopkg.in/macaron.v1"
+)
+
+type execFunc func() *httptest.ResponseRecorder
+type rateLimiterScenarioFunc func(c execFunc)
+
+func rateLimiterScenario(t *testing.T, desc string, rps int, burst int, fn rateLimiterScenarioFunc) {
+	t.Run(desc, func(t *testing.T) {
+		m := macaron.New()
+		defaultHandler := func(c *models.ReqContext) {
+			resp := make(map[string]interface{})
+			resp["message"] = "OK"
+			c.JSON(200, resp)
+		}
+		m.Use(macaron.Renderer(macaron.RenderOptions{
+			Directory: "",
+			Delims:    macaron.Delims{Left: "[[", Right: "]]"},
+		}))
+		m.Use(GetContextHandler(nil, nil, nil))
+		m.Get("/foo", RateLimit(rps, burst), defaultHandler)
+
+		fn(func() *httptest.ResponseRecorder {
+			resp := httptest.NewRecorder()
+			req, err := http.NewRequest("GET", "/foo", nil)
+			require.NoError(t, err)
+			m.ServeHTTP(resp, req)
+			return resp
+		})
+	})
+}
+
+func TestRateLimitMiddleware(t *testing.T) {
+	rateLimiterScenario(t, "rate limit calls, with burst", 10, 10, func(doReq execFunc) {
+		// first 10 calls succeed
+		for i := 0; i < 10; i++ {
+			resp := doReq()
+			assert.Equal(t, 200, resp.Code)
+		}
+
+		// next one fails
+		resp := doReq()
+		assert.Equal(t, 429, resp.Code)
+
+		// wait 1 for limiter tokens to appear, check that requests are accepted again
+		time.Sleep(1 * time.Second)
```

Comment:
Is there any way to avoid actually having to sleep in the test? It's generally not a good idea to slow down the test suite.

Reply:
This post seems to go into mocking the clocking in different ways https://dmitryfrank.com/articles/mocking_time_in_go

In general, there seems to various rate limit that support some sort of Clock type that you can advance (https://godoc.org/github.com/juju/ratelimit#NewBucketWithQuantumAndClock). So maybe worth looking at a different rate limit library?

Reply:
Thanks, I'll look into it!  either a rate limiter with injectable clock or injecting the entire rate limiter into the middleware

Reply:
Thanks, I'll look into it! I rather liked the current lib, as it's very straightforward fit and part of the "official" supplementary packages. Will look into alternatives and cosnider just injecting entire rate limiter for mocakbility as well

Reply:
I found that there's a version of `limiter.Allow` that accepts current time. So added an injectable "get current time" function param for the middleware, and got rid of `time.Sleep` in tests